README.md
# Masterblaster.py

[![Actions Status](https://img.shields.io/github/actions/workflow/status/Askepios-ai/masterblaster/build.yml?branch=beta&logo=github&style=flat-square)](https://github.com/Askepios-ai/masterblaster.py)
[![CodeCov](https://img.shields.io/codecov/c/gh/Askepios-ai/masterblaster.py/tree/beta?logo=codecov&style=flat-square)](https://app.codecov.io/gh/Askepios-ai/masterblaster.py)
[![PyPI](https://img.shields.io/pypi/v/masterblaster.py?logo=python&style=flat-square)](https://pypi.org/project/masterblaster.py)
![Docs](https://img.shields.io/readthedocs/masterblasterpy?label=Docs&link=https%3A%2F%2Fmasterblasterpy.readthedocs.io%2Fen&style=flat-square)
![DocsBeta](https://img.shields.io/readthedocs/masterblasterpy%2Fbeta?label=BetaDocs&link=https%3A%2F%2Fmasterblasterpy.readthedocs.io%2Fen%2Fbeta%2F&style=flat-square)

Python API wrapper for masterblaster.gg

## Installation

```bash
$ python -m pip install masterblaster.py
```

## Usage

Masterblaster.py is a python wrapper for the masterblaster.gg API. It allows you to easily access the data from the API and use it in your own projects.
I.e.:

```python
import asyncio
import masterblaster


async def main():
    async with masterblaster.MasterBlaster("MY-TOKEN-HERE") as m:
        for org in await m.get_all_orgs():
            for member in await org.get_members():
                print(member)


    # Alternative
    m = await masterblaster.MasterBlaster.create("MY-TOKEN-HERE")
    org = await m.get_org("MY-ORG-ID"):
    for member in await org.get_members():
        print(member)

    # Or Perhaps
    m = masterblaster.MasterBlaster("MY-TOKEN-HERE")
    ...
    ...
    async with m:
        org = await m.get_org_by_name("MY_ORG-NAME")
        for member in await org.get_members():
            print(member)


if __name__ == "__main__":
    asyncio.run(main())
```

## Contributing

Interested in contributing? Check out the contributing guidelines. 
Please note that this project is released with a Code of Conduct. 
By contributing to this project, you agree to abide by its terms.

## License

`masterblaster.py` was created by Øyvind Nohr. It is licensed under the terms
of the MIT license.

## Contributors
[![Contributors](https://img.shields.io/github/contributors/askepios-ai/masterblaster.py)](https://github.com/skepios-ai/masterblaster.py/graphs/contributors)


## Links
[Documentation](https://masterblasterpy.readthedocs.io/en/latest/)


CHANGELOG.md
# CHANGELOG



## v0.1.0-rc.8 (2023-10-25)

### Fix

* fix: mb removed allteamsready from api (#35) ([`f26187f`](https://github.com/Askepios-ai/masterblaster.py/commit/f26187f7697a82b10c03890438c727e39aad22ae))


## v0.1.0-rc.7 (2023-10-22)

### Fix

* fix: mb includes boolean for all teams ready (#34) ([`4240254`](https://github.com/Askepios-ai/masterblaster.py/commit/4240254e4154eab684af1ab7db7dceafbcc1605c))


## v0.1.0-rc.6 (2023-10-02)

### Fix

* fix: conditional parsing of datetime in matches ([`eaeb37f`](https://github.com/Askepios-ai/masterblaster.py/commit/eaeb37f69e435b4005f104d346e489961279234a))


## v0.1.0-rc.5 (2023-10-02)

### Build

* build: set masterblaster token as env-var (#32)

* build: set env-vars for testing

* test: change magic numbers for env-vars, intergration with controlled test-env on mb.gg, remove non-passing team-tests: the endpoints are intentionally left open by mb

* fix: conditional parsing of optional values ([`c5fa86c`](https://github.com/Askepios-ai/masterblaster.py/commit/c5fa86c954c40537ebc32cd4e9bb83ffa0c7f7bc))

### Chore

* chore: update readme and docs (#30)

* chore: update readme

* docs: update notebook ([`fe5d65b`](https://github.com/Askepios-ai/masterblaster.py/commit/fe5d65bb4effe9be00519804423ef20d7b84c3ea))

### Fix

* fix: secrets and variables (#33) ([`f205297`](https://github.com/Askepios-ai/masterblaster.py/commit/f205297273f32927bd75fd11e634248975fb54f4))

### Unknown

* O/dev (#31)

* build: include asyncio test dependency

* fix: masterblaster now supports(temporarily) csgo and cs2, requiring version specifiers ([`47e2540`](https://github.com/Askepios-ai/masterblaster.py/commit/47e25408f64bd8d751590057c8208bd41e599eb1))


## v0.1.0-rc.4 (2023-09-24)

### Fix

* fix: missing session-sharing between organisations and teams (#28) ([`de75d73`](https://github.com/Askepios-ai/masterblaster.py/commit/de75d73adf434b943c5c1517c90ec77f3e457a9f))


## v0.1.0-rc.3 (2023-09-23)

### Documentation

* docs: Add more badges (#24) ([`e9b88f6`](https://github.com/Askepios-ai/masterblaster.py/commit/e9b88f6eadc8824b52845aef2704ceba3dc4e49f))

### Fix

* fix: missing type (#27) ([`65bafe5`](https://github.com/Askepios-ai/masterblaster.py/commit/65bafe5af1ce65f501e968b5d9abf592140b818e))

### Unknown

* New features, test and wrappings (#25)

* build: test-env file, contributors in readme, new dependencies and pytest defaults to strict asyncio

* refactor: extending/changing responsibility in objects, less state maintained, call the api for now

* feat: team, schedule, match and organisation objects with tests. MB relates orgs to user-tokens, but evertyhing else is sentered around organisations ([`8a07dc5`](https://github.com/Askepios-ai/masterblaster.py/commit/8a07dc5de7757d10465f4aa396de3bca396bf1eb))


## v0.1.0-rc.2 (2023-09-15)

### Build

* build: publish rc and release to pypi (#18)

Co-authored-by: Askepios-ai &lt;ono016@uit.no&gt; ([`e9c66c2`](https://github.com/Askepios-ai/masterblaster.py/commit/e9c66c2238954b23032dacfff6161324d8c08eb5))

* build: poetry build before publish (#17)

Co-authored-by: Askepios-ai &lt;ono016@uit.no&gt; ([`e13fcfe`](https://github.com/Askepios-ai/masterblaster.py/commit/e13fcfe7fc945d49a3a9e78c18105a7b5a3914c5))

* build: release to pypi-indexes based on branch name (#15)

Co-authored-by: Askepios-ai &lt;ono016@uit.no&gt; ([`0b57eca`](https://github.com/Askepios-ai/masterblaster.py/commit/0b57eca7507060fb1066e8d52255e320a4438aca))

### Documentation

* docs: installation line with python (#19)

Co-authored-by: Askepios-ai &lt;ono016@uit.no&gt; ([`37a13b7`](https://github.com/Askepios-ai/masterblaster.py/commit/37a13b7ef6d806f2b6012f971bc41c1a80b3f2d1))

### Fix

* fix: versioning should not bump major, this is an unstable api (#20) ([`c7b747e`](https://github.com/Askepios-ai/masterblaster.py/commit/c7b747ed7ccdc1f56442db5acaf8714d0eb7033c))


## v0.1.0-rc.1 (2023-09-15)

### Build

* build: testing with new tokens (#14)

Co-authored-by: Askepios-ai &lt;ono016@uit.no&gt; ([`c67bd39`](https://github.com/Askepios-ai/masterblaster.py/commit/c67bd39587815d165d31bc0a31feb47f2287cf1c))

* build: use user token (#12)

Co-authored-by: Askepios-ai &lt;ono016@uit.no&gt; ([`b14b563`](https://github.com/Askepios-ai/masterblaster.py/commit/b14b5637f925f74bc43246dd7bbb90241aa2a7ca))

### Chore

* chore: updated readme with doc badges for latest/beta (#5)

* chore: updated readme with doc badges for latest/beta

* fix: remove commit artifacts ([`9798616`](https://github.com/Askepios-ai/masterblaster.py/commit/9798616996706648e6469ac83d0e95b193ab20b1))

### Documentation

* docs: update notebook ([`5050363`](https://github.com/Askepios-ai/masterblaster.py/commit/5050363d463f8f60ded6518f4f82f92fd4cd8a4a))

### Feature

* feat: initial work (#2) ([`4042161`](https://github.com/Askepios-ai/masterblaster.py/commit/404216178055d5b3ada6197a66d5133840668d3d))

### Fix

* fix: missing documentation theme (#8)

* fix: missing documentation theme

* chore: remove unused dependencies ([`51d0384`](https://github.com/Askepios-ai/masterblaster.py/commit/51d0384ae5b0ef0a874fa737aea6e1580581b2aa))

### Refactor

* refactor: poetry uploads homepage and docs to pypi, bump version (#6)

Co-authored-by: Askepios-ai &lt;ono016@uit.no&gt; ([`f605923`](https://github.com/Askepios-ai/masterblaster.py/commit/f605923653520a1832e675f5a66b032fa1ec1fdf))

### Unknown

* Change/clean deploy stage (#11) ([`1b27809`](https://github.com/Askepios-ai/masterblaster.py/commit/1b27809b74b84382f0239693700254e63a5ddc6d))

* CI Build and Release (#10)

* build(package): versioning done with semantic release

* build: CI build workflow

* build: CD test, versioning and publish workflow

Signed-off-by: Moritz Jörg &lt;14005567+mrtz-j@users.noreply.github.com&gt;

* 0.1.0-alpha.1

Automatically generated by python-semantic-release

* chore: clean up log and fix build

* chore: cleanup old release

---------


masterblaster/match.py
from datetime import datetime
from typing import Optional
from .team import Team


class MatchSettings:
    def __init__(
        self,
        useVeto: bool = None,
        useVetoSide: bool = None,
        useBestOfX: bool = None,
        mapVetoId: Optional[str] = None,
        numberOfMaps: int = None,
        isVetoCompleted: bool = None,
        isTwitchStreamingEnabled: bool = None,
        twitchStreamingUrl: Optional[str] = None,
        enableForfeitTimer: bool = None,
        timeToReadyUpBeforeForfeitInSeconds: int = None,
        scoreConfigId: Optional[str] = None,
    ) -> None:
        self.use_veto: bool = useVeto
        self.use_veto_side: bool = useVetoSide
        self.use_best_of_x: bool = useBestOfX
        self.map_veto_id: Optional[str] = mapVetoId
        self.number_of_maps: int = numberOfMaps
        self.is_veto_completed: bool = isVetoCompleted
        self.is_twitch_streaming_enabled: bool = isTwitchStreamingEnabled
        self.twitch_streaming_url: Optional[str] = twitchStreamingUrl
        self.enable_forfeit_timer: bool = enableForfeitTimer
        self.time_to_ready_up_before_forfeit_in_seconds: int = (
            timeToReadyUpBeforeForfeitInSeconds
        )
        self.score_config_id: Optional[str] = scoreConfigId


class GameSettings:
    def __init__(
        self,
        locationId: str = None,
        useAutoForfeit: bool = None,
        forfeitTimeoutInSeconds: int = None,
        customConfigFile: Optional[str] = None,
        version: int = None,
        plugins: list = None,
        insecureServer: bool = None,
    ) -> None:
        self.location_id: str = locationId
        self.use_auto_forfeit: bool = useAutoForfeit
        self.forfeit_timeout_in_seconds: int = forfeitTimeoutInSeconds
        self.custom_config_file: Optional[str] = customConfigFile
        self.version: int = version
        self.plugins: list = plugins
        self.insecure_server: bool = insecureServer


class CompetitionStatus:
    def __init__(self, status: int = 0) -> None:
        self.statuses = {
            0: "None",
            2: "Created",
            16: "Started",
            32: "Completed",
            64: "Stopped",
        }
        self.status: int = status

    def __str__(self) -> str:
        return self.statuses.get(self.status, "Unknown")

    def __repr__(self) -> str:
        return self.statuses.get(self.status, "Unknown")


class MatchStatus:
    def __init__(self, status: int = 0) -> None:
        self.statuses = {
            0: "Not started",
            10: "Starting",
            20: "Started",
            25: "Postgame",
            27: "Paused",
            28: "Failed",
            30: "Finished",
        }
        self.status: int = status

    def __str__(self) -> str:
        return self.statuses.get(self.status, "Unknown")

    def __repr__(self) -> str:
        return self.statuses.get(self.status, "Unknown")


class MatchGameServer:
    def __init__(
        self,
        serverId: str = None,
        host: str = None,
        port: int = None,
        usedAt: Optional[datetime] = None,
    ) -> None:
        self.server_id: str = serverId
        self.host: str = host
        self.port: int = port
        self.used_at: Optional[datetime] = usedAt


class MatchConnectionInfo:
    def __init__(
        self,
        host: str = None,
        port: int = None,
        paused: bool = None,
        status: int = None,
        readyAt: datetime = None,
        location: str = None,
        password: str = None,
        serverId: str = None,
        launchUrl: str = None,
        failedReason: str = None,
        launchGotvUrl: str = None,
        shutdownDelay: int = None,
        usedGameServers: list[dict] = [{}],
        connectionString: str = None,
        gotvConnectionString: str = None,
    ) -> None:
        self.host: str = host
        self.port: int = port
        self.paused: bool = paused
        self.status: int = status
        self.ready_at: Optional[datetime] = readyAt
        self.location: str = location
        self.password: str = password
        self.server_id: str = serverId
        self.launch_url: str = launchUrl
        self.failed_reason: Optional[str] = failedReason
        self.launch_gotv_url: str = launchGotvUrl
        self.shutdown_delay: Optional[int] = shutdownDelay
        self.used_game_servers: list[MatchGameServer] = [
            MatchGameServer(**server) for server in usedGameServers
        ]
        self.connection_string: str = connectionString
        self.gotv_connection_string: str = gotvConnectionString


class Match:
    """
    Class holding all information about a match

    """

    def __init__(
        self,
        competitionStatus: int,
        competitionFormat: int,
        competitionName: str,
        parentLeagueId: str,
        parentLeagueName: str,
        parentLeagueGroupId: str,
        parentLeagueSortOrder: Optional[str],
        teams: list[dict],
        series: list,
        startId: Optional[str],
        competitionId: str,
        roundIndex: int,
        matchIndex: int,
        name: str,
        game: int,
        gameId: str,
        status: int,
        statusChangedAt: Optional[datetime],
        settings: dict,
        matchType: int,
        type: int,
        format: int,
        enteredPostgameAt: Optional[datetime],
        completedAt: Optional[datetime],
        failedReason: Optional[str],
        pauseReason: Optional[str],
        pausedBy: Optional[str],
        pausedAt: Optional[datetime],
        markForDeletion: bool,
        isMatchReschedulingEnabled: bool,
        lastPossibleMatchRescheduleDate: Optional[datetime],
        matchScheduleProposals: list,
        startingAt: Optional[datetime],
        startedAt: Optional[datetime],
        createdBy: str,
        teamScores: list,
        connectionInfo: dict,
        gameSettings: dict,
        hasValidTeams: bool,
        playerIds: list,
        teamIds: list,
        hasForfeited: bool,
        hasDisqualified: bool,
        forfeitedOrDisqualified: bool,
        id: str,
        createdAt: datetime,
        allTeamsReady: bool
    ) -> None:
        self.competition_status: CompetitionStatus = CompetitionStatus(
            competitionStatus
        )
        self.compettion_format: int = competitionFormat
        self.competition_name: str = competitionName
        self.parent_league_id: str = parentLeagueId
        self.parent_league_name: str = parentLeagueName
        self.parent_league_group_id: str = parentLeagueGroupId
        self.parent_league_sort_order: Optional[str] = parentLeagueSortOrder
        self.teams = [Team(**team["team"]) for team in teams]
        self.series: list = series
        self.start_id: Optional[str] = startId
        self.competition_id: str = competitionId
        self.round_index: int = roundIndex
        self.match_index: int = matchIndex
        self.name: str = name
        self.game: int = game
        self.game_id: str = gameId
        self.status: MatchStatus = MatchStatus(status)
        self.status_changed_at: Optional[datetime] = statusChangedAt
        self.settings: MatchSettings = MatchSettings(**settings)
        self.match_type: int = matchType
        self.type: int = type
        self.format: int = format
        self.entered_postgame_at: Optional[datetime] = enteredPostgameAt
        self.completed_at: Optional[datetime] = completedAt
        self.failed_reason: Optional[str] = failedReason
        self.pause_reason: Optional[str] = pauseReason
        self.paused_by: Optional[str] = pausedBy
        self.paused_at: Optional[datetime] = pausedAt
        self.mark_for_deletion: bool = markForDeletion
        self.is_match_rescheduled: bool = isMatchReschedulingEnabled
        self.last_possible_match_reschedule_date: Optional[
            datetime
        ] = lastPossibleMatchRescheduleDate
        self.match_schedule_proposals: list = matchScheduleProposals
        self.starting_at: Optional[datetime] = startingAt
        self.started_at: Optional[datetime] = startedAt
        self.created_by: str = createdBy
        self.team_scores: list = teamScores
        self.connection_info: Optional[MatchConnectionInfo] = MatchConnectionInfo(
            **connectionInfo["fields"]
        )
        self.game_settings: GameSettings = GameSettings(**gameSettings)
        self.has_valid_teams: bool = hasValidTeams
        self.player_ids: list = playerIds
        self.team_ids: list = teamIds
        self.has_forfeited: bool = hasForfeited
        self.has_disqualified: bool = hasDisqualified
        self.forfeited_or_disqualified: bool = forfeitedOrDisqualified
        self.id: str = id
        self.created_at: datetime = createdAt
        self.all_teams_ready = allTeamsReady

    def __str__(self) -> str:
        return f"{self.name}"

    def get_teams(self) -> list[Team]:
        return self.teams

    def get_date_and_time(self) -> datetime:
        return self.starting_at

    def get_game_settings(self) -> GameSettings:
        return self.game_settings

    def get_connection_info(self) -> Optional[MatchConnectionInfo]:
        return self.connection_info

    def get_match_settings(self) -> MatchSettings:
        return self.settings

    def get_match_status(self) -> MatchStatus:
        return self.status

    def get_competition_status(self) -> CompetitionStatus:
        return self.competition_status
